flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/UnionInputGate.java

```diff
@@ -197,22 +217,23 @@ public void requestPartitions() throws IOException, InterruptedException {
 		return Optional.of(bufferOrEvent);
 	}
 
-	@Override
-	public Optional<BufferOrEvent> pollNextBufferOrEvent() throws UnsupportedOperationException {
-		throw new UnsupportedOperationException();
-	}
-
-	private InputGateWithData waitAndGetNextInputGate() throws IOException, InterruptedException {
+	private InputGateWithData waitAndGetNextInputGate(boolean blocking) throws IOException, InterruptedException {
```

Comment:
for consistency, this could now also return `Optional<...>`

Reply:
Returning optional<...> will have a bit performance penalty, still stay the same?

Reply:
Hm, if you can measure a performance penalty, I would probably not use it but you could at least add a `@Nullable` for the return value instead. However, if it is measurable, I wonder if there is not also impact from all the `Optional` use in the surrounding code.

Reply:
Try running the benchmark of `StreamNetwork*` and the results of each run are quite different, so it is difficult to measure the performance impact of `Optional`. According to my previous use, returning `Optional` in `public Optional<StreamElement> NetworkInput#pollNextElement()` has a performance penalty, because `Optional.of()` would new object. But that is per-record, and the impact should be negligible for per-buffer, so I think that adding a `@Nullable`or returning `Optional <...>` is all right.

Reply:
Couple of comments on a general level:
- if you can not measure the performance difference, then just don't bother (avoid premature optimisations) 
- make sure that you are not disturbing the benchmarks. While benchmarking, you shouldn't be touching the machine that's running the benchmarks. Scrolling window browser or changing windows (alt/cmd + tab) can seriously affect the results.
- if in doubt, verify the results more then once, like:
1. measure the base line
2. measure the change, for example +5% performance improvement
3. switch back to the base line, make sure that the result is worse
4. go back to the change and verify +5% performance improvement

I could believe that using `Optional` in `NetworkInput#pollNextElement()` might cause performance issues, but please double check that. I wouldn't be surprised that it doesn't.  `null` handling is difficult and error prone and I would like to stick with `Optional`'s compile time correctness checks as long as possible.

Reply:
> if you can not measure the performance difference, then just don't bother (avoid premature optimisations)

Okay, here I will change to return `Optional<...>`.

> I could believe that using Optional in NetworkInput#pollNextElement() might cause performance issues, but please double check that.

At testing to use `Optional` in `NetworkInput#pollNextElement()`, I did as you said.

Reply:
The updated code has changed `waitAndGetNextInputGate()` to return `Optional<...>`.